### PR TITLE
Bk/fix initial attributes target content offset

### PR DESCRIPTION
**Pull request by bryankeller**
## Details

This PR fixes 1 issue (can't find a fix for the other, sadly):

1. In certain cases, the initial layout attributes of an inserting item would be in the wrong spot. This would happen when the target content offset adjustment was changing the position that the newly inserted item would appear at. We need to compensate for this.

Before:

https://github.com/airbnb/MagazineLayout/assets/746571/e1d9c2b1-b7d3-4f35-a030-2d6c5845936a

After:

https://github.com/airbnb/MagazineLayout/assets/746571/b415ac02-c1c3-4e63-b14f-8b939a29729c

-----

2. **[Not fixed, seems like a collection view bug where it skips trying to animate an item since it's end position is off screen]** There's an off edge case with collection view that causes a moved item to render instantly at its final position, rather than smoothly sliding over. It's related to self-sizing cells (doesn't happen for statically sized items). All returned layout attributes look correct, so I don't think it's `MagazineLayout`'s fault. It happens when the inserted item / moved item is close to the the top of the screen, which makes me think internally, `UICollectionView` thinks the moving item isn't visible, and therefore short-circuits its rendering. Interestingly, reducing the estimated item height from `150` -> `10` fixes it. 150 was a magic number, and 10 is too, but I think using a smaller number makes it less likely for collection view to hit this issue.

Before:

https://github.com/airbnb/MagazineLayout/assets/746571/3f0df00a-4463-4916-add9-6a140dc4c665

## Related Issue

N/A

## Motivation and Context

Bug fixes

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: MagazineLayout/Public/MagazineLayout.swift [hunk 1]
```diff
@@ -459,7 +465,7 @@ public final class MagazineLayout: UICollectionViewLayout {
     {
       return previousLayoutAttributesForItem(at: initialIndexPath)
     } else {
-      return super.layoutAttributesForItem(at: itemIndexPath)
```
Comment (bryankeller) [inline MagazineLayout/Public/MagazineLayout.swift:462]: noticed this was the wrong super call

File: MagazineLayout/Public/MagazineLayout.swift [hunk 2]
```diff
@@ -490,7 +496,7 @@ public final class MagazineLayout: UICollectionViewLayout {
       itemLayoutAttributesForPendingAnimations[finalIndexPath] = attributes
       return attributes
     } else {
-      return super.layoutAttributesForItem(at: itemIndexPath)
```
Comment (bryankeller) [inline MagazineLayout/Public/MagazineLayout.swift:493]: noticed this was the wrong super call

File: MagazineLayout/Public/MagazineLayout.swift [hunk 3]
```diff
@@ -755,6 +763,7 @@ public final class MagazineLayout: UICollectionViewLayout {
   override public func invalidateLayout(with context: UICollectionViewLayoutInvalidationContext) {
     guard let context = context as? MagazineLayoutInvalidationContext else {
       assertionFailure("`context` must be an instance of `MagazineLayoutInvalidationContext`")
+      super.invalidateLayout(with: context)
```
Comment (bryankeller) [inline MagazineLayout/Public/MagazineLayout.swift:766]: noticed this was missing a call to super, although this assert / guard should never be hit

File: MagazineLayout/Public/MagazineLayout.swift [hunk 4]
```diff
@@ -862,13 +875,6 @@ public final class MagazineLayout: UICollectionViewLayout {
   // Used to provide the model state with the current visible bounds for the sole purpose of
   // supporting pinned headers and footers.
   private var currentVisibleBounds: CGRect {
-    let contentInset: UIEdgeInsets
-    if #available(iOS 11.0, tvOS 11.0, *) {
-      contentInset = currentCollectionView.adjustedContentInset
-    } else {
-      contentInset = currentCollectionView.contentInset
-    }
-
```
Comment (bryankeller) [inline MagazineLayout/Public/MagazineLayout.swift:871]: not needed - I added a computed property for `contentInset` that does this same logic, so we can use that directly

File: MagazineLayout/Public/Types/MagazineLayoutItemSizeMode.swift [hunk 5]
```diff
@@ -132,7 +132,7 @@ public enum MagazineLayoutItemHeightMode: Hashable {
   /// For example, if you support multiline labels or dynamic type, your height is likely not known
   /// until the Auto Layout engine resolves the layout at runtime.
   public static var dynamic: MagazineLayoutItemHeightMode {
-    .dynamic(estimatedHeight: MagazineLayout.Default.ItemHeight)
+    .dynamic(estimatedHeight: 10)
```
Comment (bryankeller) [inline MagazineLayout/Public/Types/MagazineLayoutItemSizeMode.swift:135]: I need to do some testing to make sure this doesn't break anything in the Airbnb app

File: MagazineLayout/Public/MagazineLayout.swift (hunk 2 above)
Comment (brynbodayle) [reply MagazineLayout/Public/MagazineLayout.swift:493]: oh wow, were these ever getting hit though?

File: MagazineLayout/Public/MagazineLayout.swift [hunk 6]
```diff
@@ -862,6 +873,7 @@ public final class MagazineLayout: UICollectionViewLayout {
   // animations will start from the estimated height.
   private var itemLayoutAttributesForPendingAnimations = [IndexPath: UICollectionViewLayoutAttributes]()
   private var supplementaryViewLayoutAttributesForPendingAnimations = [IndexPath: UICollectionViewLayoutAttributes]()
+  private var targetContentOffsetCompensatingYOffsetForAppearingItem: CGFloat?
```
Comment (brynbodayle) [inline MagazineLayout/Public/MagazineLayout.swift:876]: could we document this?

File: MagazineLayout/Public/Types/MagazineLayoutItemSizeMode.swift [hunk 7]
```diff
@@ -132,7 +132,7 @@ public enum MagazineLayoutItemHeightMode: Hashable {
   /// For example, if you support multiline labels or dynamic type, your height is likely not known
   /// until the Auto Layout engine resolves the layout at runtime.
   public static var dynamic: MagazineLayoutItemHeightMode {
-    .dynamic(estimatedHeight: MagazineLayout.Default.ItemHeight)
```
Comment (brynbodayle) [inline MagazineLayout/Public/Types/MagazineLayoutItemSizeMode.swift:135]: See we still use this constant elsewhere, should we change the constant value instead? A bit confused on when we want this smaller height versus not.

File: MagazineLayout/Public/Types/MagazineLayoutItemSizeMode.swift (hunk 5 above)
Comment (brynbodayle) [reply MagazineLayout/Public/Types/MagazineLayoutItemSizeMode.swift:135]: Hmm, not sure about this. I think a lower estimated height could be mean more cells are initially loaded, it's a less accurate estimation. So we could be receiving worse performance to address the animation edge case.

Does the issue only occur if the estimated height puts the original layout attributes frame intersecting the top of the collection view?

File: MagazineLayout/Public/MagazineLayout.swift (hunk 2 above)
Comment (bryankeller) [reply MagazineLayout/Public/MagazineLayout.swift:493]: Yeah I haven't seen it hit

File: MagazineLayout/Public/Types/MagazineLayoutItemSizeMode.swift (hunk 5 above)
Comment (bryankeller) [reply MagazineLayout/Public/Types/MagazineLayoutItemSizeMode.swift:135]: Those are good points - I'll do some more investigation

File: MagazineLayout/Public/Types/MagazineLayoutItemSizeMode.swift (hunk 7 above)
Comment (bryankeller) [reply MagazineLayout/Public/Types/MagazineLayoutItemSizeMode.swift:135]: I realized we were using this default item height for not just the estimated height, but also the default static height if someone doesn't implement `sizeModeForItemAt`. Didn't want to cause a regression in that case.

**Review by brynbodayle — APPROVED** [review]

**Pull request merged**